runtime/libia2/include/permissive_mode.h

```diff
@@ -465,23 +465,78 @@ __attribute__((constructor)) void permissive_mode_init(void) {
   install_permissive_mode_handler();
 }
 
-/*
- * Constructor to wait for the logging thread to finish and log the memory map.
- * If a process forks and execve's this function will not be called.
- */
-__attribute((destructor)) void wait_logging_thread(void) {
-  exiting = true;
-  pthread_join(logging_thread, NULL);
+// `getline` calls `malloc` inside of `libc`,
+// but we wrap `malloc` with `__wrap_malloc`,
+// so we need to free what `getline` allocated with `__real_free`.
+void __real_free(void *ptr);
+
+extern uintptr_t ia2_stack_addrs[IA2_MAX_COMPARTMENTS];
+
+void log_memory_map(void) {
   FILE *log = fopen(log_name, "a");
   assert(log);
   FILE *maps = fopen("/proc/self/maps", "r");
   assert(maps);
-  char tmp[256] = {0};
-  while (fgets(tmp, sizeof(tmp), maps)) {
-    fprintf(log, "%s", tmp);
-    memset(tmp, 0, sizeof(tmp));
+
+  // Skip dev and inode.
+  fprintf(log, "  start addr-end addr     perms offset  path\n");
+
+  char *line = NULL;
+  size_t line_cap = 0;
+  while (true) {
+    const ssize_t line_len = getline(&line, &line_cap, maps);
+    if (line_len == -1) {
+      break;
+    }
+
+    // Remove trailing newline.
+    if (line_len > 0 && line[line_len - 1] == '\n') {
+      line[line_len - 1] = 0;
+    }
+
+    // Parse `/proc/self/maps` line.
+    uintptr_t start_addr = 0;
+    uintptr_t end_addr = 0;
+    char perms[4] = {0};
+    uintptr_t offset = 0;
+    unsigned int dev_major = 0;
+    unsigned int dev_minor = 0;
+    ino_t inode = 0;
+    int path_index = 0;
+    sscanf(line, "%lx-%lx %4c %lx %x:%x %lu %n", &start_addr, &end_addr, perms, &offset, &dev_major, &dev_minor, &inode, &path_index);
```

Comment:
We should warn if this fails rather than printing bogus info.

Reply:
Fixed in #579.